src/transformers/models/imagebind/modeling_imagebind.py

```diff
@@ -0,0 +1,2062 @@
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+    ) -> torch.Tensor:
+        seq_length = input_ids.shape[-1] if input_ids is not None else inputs_embeds.shape[-2]
+
+        if position_ids is None:
+            position_ids = self.position_ids[:, :seq_length]
+
+        if inputs_embeds is None:
+            inputs_embeds = self.token_embedding(input_ids)
+
+        position_embeddings = self.position_embedding(position_ids)
+        embeddings = inputs_embeds + position_embeddings
+
+        return embeddings
+
+
+class ImageBindAttention(nn.Module):
+    """Multi-headed attention from 'Attention Is All You Need' paper. This is [`CLIPAttention`] with key and value biases"""
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.embed_dim = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.embed_dim // self.num_heads
+        if self.head_dim * self.num_heads != self.embed_dim:
+            raise ValueError(
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim} and `num_heads`:"
+                f" {self.num_heads})."
+            )
+        self.scale = self.head_dim**-0.5
+        self.dropout = config.attention_dropout
+
+        self.qkv_proj = nn.Linear(self.embed_dim, self.embed_dim * 3)
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim)
+
+        self.add_kv_bias = config.add_kv_bias
+
+        # Create bias parameters for key and value sequences.
+        if self.add_kv_bias:
+            self.k_bias = nn.Parameter(torch.empty((1, 1, self.embed_dim)))
+            self.v_bias = nn.Parameter(torch.empty((1, 1, self.embed_dim)))
+        else:
+            self.k_bias = None
+            self.v_bias = None
+
+    def _shape(self, tensor: torch.Tensor, seq_len: int, batch_size: int):
+        return tensor.view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2).contiguous()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        """Input shape: Batch x Time x Channel"""
+
+        batch_size, seq_len, embed_dim = hidden_states.size()
+
+        qkv = self.qkv_proj(hidden_states).reshape(batch_size, seq_len, 3, -1).permute(2, 0, 1, 3)
+        query_states, key_states, value_states = qkv.unbind(0)
+
+        query_states = query_states * self.scale
+
+        # Add key/value biases if necessary
+        if self.add_kv_bias:
+            # Repeat bias along batch dimension (first)
+            key_states = torch.cat([key_states, self.k_bias.repeat(batch_size, 1, 1)], dim=1)
+            value_states = torch.cat([value_states, self.v_bias.repeat(batch_size, 1, 1)], dim=1)
```

Comment:
Because that's how it was done in the original repo so it was just reproduced the same in the modeling file as well:
https://github.com/facebookresearch/ImageBind/blob/3fcf5c9039de97f6ff5528ee4a9dce903c5979b3/imagebind/models/transformer.py#L40 

Although I can make the changes if needed and check if the test passes and the precision of outputs is still the same

Reply:
Update: Okay I changed the code to the following to make it more transformers compliant:
```
        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim)
        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim)
        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim)
        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim)

        # Create bias parameters for key and value sequences.
        if self.add_kv_bias:
            self.k_bias = nn.Parameter(torch.empty((1, 1, self.embed_dim)))
            self.v_bias = nn.Parameter(torch.empty((1, 1, self.embed_dim)))
        else:
            self.k_bias = None
            self.v_bias = None

    def _shape(self, tensor: torch.Tensor, seq_len: int, bsz: int):
        return tensor.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2).contiguous()

    def forward(
        self,
        hidden_states: torch.Tensor,
        attention_mask: Optional[torch.Tensor] = None,
        causal_attention_mask: Optional[torch.Tensor] = None,
        output_attentions: Optional[bool] = False,
    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
        """Input shape: Batch x Time x Channel"""

        bsz, tgt_len, embed_dim = hidden_states.size()

        key_states = self.k_proj(hidden_states)
        value_states = self.v_proj(hidden_states)

        # get query proj
        query_states = self.q_proj(hidden_states) * self.scale

        # Add key/value biases if necessary
        if self.add_kv_bias:
            # Repeat bias along batch dimension (first)
            key_states = torch.cat([key_states, self.k_bias.repeat(bsz, 1, 1)], dim=1)
            value_states = torch.cat([value_states, self.v_bias.repeat(bsz, 1, 1)], dim=1)

        key_states = self._shape(key_states, -1, bsz)
        value_states = self._shape(value_states, -1, bsz)

```
Things it breaks:
- modeling test failure:
```
===================================== short test summary info ======================================
FAILED tests/models/imagebind/test_modeling_imagebind.py::ImageBindTextModelTest::test_initialization - AssertionError: 0.000663497019559145 not found in [0.0, 1.0] : Parameter text_model.encoder.lay...
FAILED tests/models/imagebind/test_modeling_imagebind.py::ImageBindVisionModelTest::test_initialization - AssertionError: 0.0068885041400790215 not found in [0.0, 1.0] : Parameter vision_model.encoder....
FAILED tests/models/imagebind/test_modeling_imagebind.py::ImageBindAudioModelTest::test_initialization - AssertionError: -0.00402283389121294 not found in [0.0, 1.0] : Parameter audio_model.encoder.la...
FAILED tests/models/imagebind/test_modeling_imagebind.py::ImageBindModelTest::test_initialization - AssertionError: -0.004465410951524973 not found in [0.0, 1.0] : Parameter text_model.encoder.la...
```
- `ImageBindModelIntegrationTest` failure due to change in outputs:
```
FAILED tests/models/imagebind/test_modeling_imagebind.py::ImageBindModelIntegrationTest::test_inference - AssertionError: False is not true
```
- Due to removal of `qkv_proj` and addition of `k_proj`, etc. the model parameter has to be changed and also the weight conversion file

Reply:
Yes indeed the weights conversion file has to be modified and reuploaded. We try to have separated q k v, only a minority of models have qkv fused.. I'll let @ArthurZucker decide on that one but I'm leaning on the "update converter and weights" side. Then, the results should be the same w.r.t. integration tests.

Reply:
Usually splitting is asked see #33928 is it basically enables cheap finetuning! 🤗 Let's go with this

Reply:
this would also standardize this attention layer as it can be simply copied from Gemma!